AI 2008 fld2 and PathFinding algorithm

Henrybk opened the pull request:
As some of you may know I have been working on writting a new pathfinding code for openkore for quite a while, I believe today I finally got it working and decided to PR it so other developers can take a look.

Features:
```
1 - Uses new simplified fld2 files
2 - Uses .weight files based on .dist files instead of .dist files, allowing custom static or dynamic weight maps (avoid mvps, non-walking mobs, dangerous map areas)
3 - Correctly uses diagonals, using weight 1.4 that of ortogonal moving
4 - Uses diagonal distance heuristic, which is the most appropriate for ragnarok maps
5 - Is guaranteed to always find the shortest / best path since it does not limit openlist
6 - Uses about 42% less time to calculate a route
7- Fixes runfromTarget, attackCheckLOS and AttackCanSnipe
8- Corretly predicts the time to walk a given distance, preventing a few timeouts
```

benchmark of current algorithm: https://gyazo.com/cd3adf01a6af0b022e5e57a4010c7064

benchmark of my algorithm: https://gyazo.com/2359320cab8587a1e9078558b99ac4d9

My benchmark consists of all the portalslos paths plus 20 randomwalks on all RO maps.

Edit: Most of the changes were ideas from here: http://forums.openkore.com/viewtopic.php?f=36&t=1350

this pull fixes #2546

c4c1n6kr3m1 commented:
my problem untill now is all bots walk on the same path, so obvious, only 0 or 1 for weight
need better option for that

Henrybk commented:
That’s the idea with custom dynamic weight maps, it would be pretty easy to write a plugin to make bots walk on diferent but short paths, just add randomly 5~15 weight to all walkable cells on the mao when the map is loaded

c4c1n6kr3m1 commented:
this is awesome,  
hope you give example for the plugin
Random path weight and better Line of Sight  will make better bot

Henrybk commented:
Just made some new benchmarks and found that my code is almost twice as fast

c4c1n6kr3m1 commented:
just tried this "Remake-routing" branch

first run with win7 32 bit
1st error
crash when compiling portals
2nd error
```Error message:
Undefined subroutine &Utils::makeWeightMap called at src/Field.pm line 336.
```
@Henrybk ???

Henrybk commented:
@c4c1n6kr3m1 I can't find a way to compile XSTools in windows, so the code will only work on linux OS for now

Xmpeshka commented:
Getting an error when compiling portals,  

> Can't call method "isWalkable" on an undefined value at src/Misc.pm line 772.

Henrybk commented:
This PR has been updated and should now be working if properly compiled.
I have also updated and compressed all fld2 files.

Henrybk commented on src/Misc.pm:
```diff
@@ -4308,7 +4308,7 @@ sub checkSelfCondition {
 	if ($config{$prefix."_whenWater"}) {
 		my $pos = calcPosition($char);
-		return 0 if ($field->getBlock($pos->{x}, $pos->{y}) != Field::WALKABLE_WATER);
+		return 0 if ($field->isWater($pos->{x}, $pos->{y}));
```
This should be a negation

Henrybk commented on src/Task/Route.pm:
```diff
@@ -318,7 +318,7 @@ sub iterate {
 				# for example.
 				my %nextPos = (x => $self->{new_x}, y => $self->{new_y});
-				if (distance(\%nextPos, $pos) > $config{$self->{actor}{configPrefix}.'route_step'}) {
+				if (distance(\%nextPos, $pos) > ($config{$self->{actor}{configPrefix}.'route_step'} * 2)) {
```
This change should be reviewed, the same check is also made in maproute and seems to be able to cause some errors

c4c1n6kr3m1 replied:
now, i think i understand why my old bot frequently stuck while on stairs
almost arrived on destination on stair , but move to different floor to attack monster
after monster is killed, stuck, 
because , although very short distance but on different floor, the LOS from last point should check too?

Xmpeshka commented:
Help =) Can u compile xstools for your bild. I tried compile trying compile it by get errors
`src\auto\XSTools\PathFinding\algorithm.cpp
In file included from c:\mingw\lib\gcc\mingw32\6.3.0\include\c++\cstdlib:75:0,
                 from c:\mingw\lib\gcc\mingw32\6.3.0\include\c++\stdlib.h:36,
                 from src\auto\XSTools\PathFinding\algorithm.cpp:1:
c:\mingw\include\stdlib.h:145:115: error: 'uintptr_t' has not been declared
   typedef void (__cdecl *_invalid_parameter_handler)(const wchar_t *,const wchar_t *,const wchar_t *,unsigned int,uintp
tr_t);
                                                                                                                   ^~~~~
~~~~
src\auto\XSTools\PathFinding\algorithm.cpp: In function 'int CalcPath_pathStep(CalcPath_session*)':
src\auto\XSTools\PathFinding\algorithm.cpp:178:6: warning: unused variable 'indexNeighbor' [-Wunused-variable]
  int indexNeighbor = 0;
      ^~~~~~~~~~~~~
src\auto\XSTools\PathFinding\algorithm.cpp:179:6: warning: unused variable 'nodeList' [-Wunused-variable]
  int nodeList;
      ^~~~~~~~
scons: *** [src\auto\XSTools\PathFinding\algorithm.os] Error 1
scons: building terminated because of errors.
-e "\e[1;31mCompilation failed. Please read http://wiki.openkore.com/index.php/How_to_run_OpenKore for help.\e[0m"`

JpnHub commented:
> 2 - Uses .weight files based on .dist files instead of .dist files, allowing custom static or dynamic weight maps (avoid mvps, non-walking mobs, dangerous map areas)

Would it be possible to create a small and simple plugin to showcase the dynamic real time weight modifications? 

As from my current testing, everything seems to be an overall huge improvement in both speed and efficiency, but the dynamic modification of the weight system is  a tad bit confusing to understand from your code alone, which is a shame considering the immense improvements it could bring alongside with itself.

Henrybk commented:
> > 2 - Uses .weight files based on .dist files instead of .dist files, allowing custom static or dynamic weight maps (avoid mvps, non-walking mobs, dangerous map areas)
> 
> Would it be possible to create a small and simple plugin to showcase the dynamic real time weight modifications?
> 
> As from my current testing, everything seems to be an overall huge improvement in both speed and efficiency, but the dynamic modification of the weight system is a tad bit confusing to understand from your code alone, which is a shame considering the immense improvements it could bring alongside with itself.

So, the branch which was used for this PR is able of dynamic path modification, but not efficiently, you would need to recalculate all the remaining path to the destination.
This branch uses A star for its calculations, I have 3 others branches of this one which improve exactly on the dynamic pathing, one uses Lifelong Planing A star, another one uses D star Lite and the third is a incomplete unfinished version of a Moving Target D star Lite implementation.

On the D star Lite branch I use a plugin (https://github.com/Henrybk/openkore/blob/D_star_lite_field_cell_increase/plugins/DStarLiteAvoid.pl) to add obstacles to the weight map, which get resolved by Task::Route, which makes Task::Pathfinding recalculate only the steps necessary for the obstacle to be avoided.

Here is a gif of this method working to avoid a monster: https://i.gyazo.com/a37fbb7b59a16099c9b9b30df0f51bc5.mp4

eprabotmesmo commented:
Estou querendo testar esse patch mais está dando problema para compilar os portais se eu tirar todos os mapas que dão problema ele funciona @Henrybk  tem ideia do que possa está ocasionando o erro?
os mapas que lembro de cabeça que dão erro são "har_in01, rockrdg1, juperos_01, lighthalzen".

obs.: Já tentei extrair os mesmo é converter para fl2 não deu certo.
obs2.: No linux o problema não existe.

![image](https://user-images.githubusercontent.com/44814993/51886801-14930700-2370-11e9-976f-51420da5da57.png)

Henrybk commented:
Updated branch to master, fixed all remaining known bugs, made a few changes to make it about 5% faster, made a lot of code cleaning and added some comments.

Henrybk commented:
I also have a branch in my fork in which I applied @lututui PR #2402 so I can compile it in windows.
Branch: https://github.com/Henrybk/openkore/tree/Routing-clen%2Bbuild

Xmpeshka commented:
Thank u, Still waiting a windows bulid "_*

Xmpeshka commented:
![image](https://user-images.githubusercontent.com/42699011/54980120-f800fd00-4fb5-11e9-8683-86d54d33d097.png)
Crashes on compiling portals

Henrybk commented:
@Xmpeshka Did you compile xstools properly? You must have
`Python - 2.7.1 (x86)
Perl Strawberry - 5.28 (x86)
C compiler`
while using the branch with lututui's build changes and use `gmake` while on the openkore root directory in cmd/powershell.

Otherwise if you want to test just just install Perl Strawberry - 5.28 (x86) and I'll compile xstools and add the compiled dlls to the branch.

Xmpeshka commented:
same situation on another pc with yours precompiled dlls strawbery x86 5.28.1.1

on pc where i compiling myself i use 5.28.1.1+python+minigw

Henrybk commented:
Strange, I just compiled portalsLOS here without issues. Could you please translate the error message you are getting?

Xmpeshka commented:
Standard windows error 
![image](https://user-images.githubusercontent.com/42699011/55011456-f1df4080-4ff6-11e9-99e0-a15b52321a71.png) like this
Starting openkore - using openkore.pl

Xmpeshka commented:
deleted strings aldeg_cas** from portals.txt and portals compiled normaly

Henrybk commented:
Can you provide the portals.txt you are using? I can compile aldeg_cas05 normally. Also, just to be sure, are you using openkore.pl instead of the .exe's?

Xmpeshka commented:
using portals from branch, server -localhost for test, launching openkore.pl

Henrybk commented:
thanks for testing, I'll check into that error

Henrybk commented:
Activating debug doesn't show anything in the console?

Xmpeshka commented:
new problem - no moving: attack, west 5, randomRouting - Char stays on ground like a stone =)

Xmpeshka commented:
> Activating debug doesn't show anything in the console?

![image](https://user-images.githubusercontent.com/42699011/55016941-cfeabb80-5000-11e9-9d3e-564bc54a9064.png) then crashes

Henrybk commented:
> new problem - no moving: attack, west 5, randomRouting - Char stays on ground like a stone =)

Fixed this in the last commit in both branches, I had forgotten to invert the order of the array of coordinates returned to Route.pm from Pathfinding.pm

Xmpeshka commented:
Compiled new dll:
west east comands and, attack with move to target - works
move x y, randomroute - don't work

Henrybk commented:
@Xmpeshka Fixed the map compilation crash, I had not implemented a failsafe for when the user tries to solve a path with invalid coordinates.

Turns out the portal
`aldeg_cas05 188 49 aldeg_cas05 165 288`
cannot exist, since the map is only 240x240.

Xmpeshka commented:
> @Xmpeshka Fixed the map compilation crash, I had not implemented a failsafe for when the user tries to solve a path with invalid coordinates.
> 
> Turns out the portal
> `aldeg_cas05 188 49 aldeg_cas05 165 288`
> cannot exist, since the map is only 240x240.
> 
> Edit: also I was able to use move x y and randomWalk without any issues.

My fault -  forgot about route_Step 7 and used my own conf files
Working normal - but little bug  on my conf
 i have a eventMacro that talks in noob zone
```
do move 55 110
	do ai manual
	do move 55 110
	do move 55 110
	do talknpc 53 114 c c c c c c c c n
	pause &rand(2,5)
	do move 147 112
	pause &rand(5,10)
	do move 100 25
        ....
```
Kore don't talk with npc and walks next
`[Mar 27 22:42:14 2019.50] AI set to manual mode
[Mar 27 22:42:15 2019.57] Move You - (re)trying
[Mar 27 22:42:15 2019.69] You're moving from (53, 111) to (55, 110) - distance 2.2
[Mar 27 22:42:16 2019.12] Move You - (re)trying
[Mar 27 22:42:16 2019.23] You're moving from (55, 110) to (55, 110) - distance 0.0
[Mar 27 22:42:16 2019.23] Move You - done
[Mar 27 22:42:16 2019.62] new_2-1 148 112=new_2-2 100 9, 113
[Mar 27 22:42:16 2019.62] new_2-2 100 6=new_2-1 144 112, 136
[Mar 27 22:42:16 2019.63] Distance to portal (new_2-1 148 112=new_2-2 100 9) is 93
[Mar 27 22:42:16 2019.63] Attempting to teleport near portal, try #1
[Mar 27 22:42:16 2019.64] You don't have the Teleport skill or a Fly Wing
[Mar 27 22:42:16 2019.70] Move You - (re)trying
[Mar 27 22:42:16 2019.81] You're moving from (55, 110) to (63, 111) - distance 8.1
[Mar 27 22:42:17 2019.24] Move You - (re)trying
[Mar 27 22:42:17 2019.30] You're moving from (58, 111) to (63, 111) - distance 5.0
[Mar 27 22:42:17 2019.30] Move You - done
[Mar 27 22:42:17 2019.31] Move You - (re)trying
[Mar 27 22:42:17 2019.84] Move You - (re)trying
[Mar 27 22:42:17 2019.90] You're moving from (62, 111) to (67, 111) - distance 5.0
[Mar 27 22:42:18 2019.33] Move You - timeout
`

Henrybk commented:
Thanks again for testing, I will take a look at that when I ger the time

Henrybk commented:
I just tested it with you macro and was able to talk to sprakki normally

https://i.gyazo.com/ba0c3ba35725ef8f47a8dd9c8aded849.png

Xmpeshka commented:
hmm it's called from another automacro - but i copied by all folders from my main kore excluding src - nothing changes in config except routeStep if at this moment manualy wrote in console move 55 110 - it's talk. same problem with all novice ground - kore skips talking with an npcs
![image](https://user-images.githubusercontent.com/42699011/55112213-161d4900-50ed-11e9-8a1d-3dc0419f9754.png)
and finaly stopts in this point 126 106
![image](https://user-images.githubusercontent.com/42699011/55112274-482eab00-50ed-11e9-95ca-05449c26f192.png)
`
------------------ AI Sequence -------------------
0: route

1: eventMacro {mapChanged=>'1553720616.04922'}

--------------------------------------------------
`

eprabotmesmo commented:
![image](https://user-images.githubusercontent.com/44814993/55159158-3544c680-513f-11e9-99b9-0a82250e9f76.png)

Henrybk commented:
@eprabotmesmo
Firstly, did you use the compiled branch ou did you compile it yourself? Are you using your own config files or the ones on my branch? Did you use openkore.pl? If so, can you post another test with debug on?

Xmpeshka commented:
also tested thief quests - macroses works normaly. Bug only with newbie zone. Also strange attemps to teleport from portal at the start of newzone - there is one npc and bridge =) no portals etc - standart zone of official revoserver

Xmpeshka commented:
![image](https://user-images.githubusercontent.com/42699011/55172452-2257e380-518b-11e9-8294-6329cd3d7fba.png)
also some little bug )

eprabotmesmo commented:
> @eprabotmesmo
> Firstly, did you use the compiled branch or did you compile it yourself? Are you using your own config files or the ones on my branch? Did you use openkore.pl? If so, can you post another test with debug on?
I using your branch files
![image](https://user-images.githubusercontent.com/44814993/55176056-2c63ed00-515f-11e9-8c1e-bb7a396f3e5c.png)

sctnightcore commented:
@Xmpeshka  test with openkore not icekore / windkore !

Xmpeshka commented:
> @Xmpeshka test with openkore not icekore / windkore !

waht difference ? i take openkore-Routing-clen-build - and changed only network packets tables and injection part - all other src are from this branch

sctnightcore commented:
@Xmpeshka  I think it's different.

Xmpeshka commented:
> @Xmpeshka I think it's different.

u can test it on windkore - it will be a good expirience - maybe i doing something wrong

Xmpeshka commented:
Tested with the last master branch - all ok with macros.

Henrybk commented:
> Tested with the last master branch(with network changes affects my server) - all ok with macros.

Thanks again for testing.

Henrybk commented:
> > @eprabotmesmo
> > Firstly, did you use the compiled branch or did you compile it yourself? Are you using your own config files or the ones on my branch? Did you use openkore.pl? If so, can you post another test with debug on?
> > I using your branch files
> > ![image](https://user-images.githubusercontent.com/44814993/55176056-2c63ed00-515f-11e9-8c1e-bb7a396f3e5c.png)

I'll take a look at that, thanks for testing and reporting.

Henrybk commented:
> ![image](https://user-images.githubusercontent.com/42699011/55172452-2257e380-518b-11e9-8294-6329cd3d7fba.png)
> also some little bug ) w taking loot



> > @eprabotmesmo
> > Firstly, did you use the compiled branch or did you compile it yourself? Are you using your own config files or the ones on my branch? Did you use openkore.pl? If so, can you post another test with debug on?
> > I using your branch files
> > ![image](https://user-images.githubusercontent.com/44814993/55176056-2c63ed00-515f-11e9-8c1e-bb7a396f3e5c.png)

Last commit should have fixed both of those issues, I updated both branches.

Henrybk commented:
These last changes should be the last ones unless someone finds any more bugs.
In them I updated to master, which now uses lututui's build changes branch, so no need anymore of using the other branch to test. I also changed the base weight of any cell to 0, so it won't affect the best route calculation (which uses 10 for ortogonal and 14 for diagonal moves), walls now use weight -1.
I also commented some error messages because they would be too common since openkore calls a lot of dumb pathfinding requests.

Xmpeshka commented:
https://youtu.be/JTBGceRjcQY
i recorded some video - with that macro bus

Nipodemos commented on control/config.txt:
```diff
@@ -397,6 +399,7 @@ homunculus_attackMaxDistance 2.5
 homunculus_attackMaxRouteTime 4
 homunculus_attackCanSnipe 0
 homunculus_attackCheckLOS 0
+mercenary_attackAdjustLOSMaxRouteDistance 10
```
typo:
should be homunculus here

Xmpeshka commented:
![image](https://user-images.githubusercontent.com/42699011/62855828-e9647980-bcfb-11e9-904a-567ca57e3776.png)

Glikadin commented:
Hello. Today i am testing remake-routing.
Have meny problems:

Test #1
```
move 167 207 izlude
move 175 207 izlude
move 63 97 izlude_in
move 65 121 izlude_in
```
Result:
![try1](https://user-images.githubusercontent.com/20753532/62920512-a9aa9a00-bdae-11e9-8b29-247e1c31291d.jpg)

The bot went to an unpredictable point on the map and got stuck there.

Glikadin commented:
Hello. Today i am testing remake-routing.

After i fixed 

> routeweights.txt
```
PORTAL 100
```
The bot finds the way well
But have problem:
During movement, the bot constantly pauses (((
Video of my testing:
https://youtu.be/MWEnuPa0D0U

Glikadin commented:
@Henrybk Please continue work

Henrybk commented:
@Glikadin oh, thanks for testing, I`ll take a look this weekend.

Henrybk commented:
Please continue in #2871
